### PR TITLE
CATROID-88 Devices without Flash should ignore FlashBrick

### PR DESCRIPTION
**Pull request by DinosaurierRex**
When camera device has no flash available, the flash brick action should simply be ignored.
e.g. emulators or tablets
Added tests for FlashUtil so CameraManager and Camera are PowerMocks for simulating environment.

### CODE REVIEW

File: catroid/src/main/java/org/catrobat/catroid/camera/CameraManager.java [hunk 1]
```diff
@@ -96,8 +96,14 @@ public static CameraManager getInstance() {
 		return instance;
 	}
 
+	public static void setInstance(CameraManager cameraManager) {
+		instance = cameraManager;
+	}
+
 	public static void makeInstance() {
-		instance = new CameraManager();
+		if (instance == null) {
```
Comment (rschluesselbauer) [inline catroid/src/main/java/org/catrobat/catroid/camera/CameraManager.java]: I agree with this, because the constructor does some work & it's unnecessary to do it for each camera permission.
There are tests that possibly rely on makeInstance() resetting the CameraManager.
I have not had time to look into it further, but it's possible that the Stage too relies on CameraManager having a clean state. Check this out and see if it's worth it!

File: catroid/src/main/java/org/catrobat/catroid/stage/StageResourceHolder.java [hunk 2]
```diff
@@ -489,10 +486,11 @@ private void showResourceFailedErrorDialog() {
 					failedResourcesMessage = failedResourcesMessage + stageActivity.getString(R.string
 							.prestage_no_front_camera_available);
 					break;
-				case Brick.CAMERA_FLASH:
-					failedResourcesMessage = failedResourcesMessage + stageActivity.getString(R.string
-							.prestage_no_flash_available);
-					break;
+				// CATROID-88: Flash brick should be ignored
```
Comment (rschluesselbauer) [inline catroid/src/main/java/org/catrobat/catroid/stage/StageResourceHolder.java]: Please remove commented out code.

File: catroid/src/main/java/org/catrobat/catroid/utils/FlashUtil.java [hunk 3]
```diff
@@ -112,6 +124,11 @@ public static void initializeFlash() {
 
 	public static void flashOn() {
 		Log.d(TAG, "flashOn()");
+
+		if (!isAvailable()) {
+			return;
+		}
+
 		if (!CameraManager.getInstance().hasCurrentCameraFlash()) {
```
Comment (rschluesselbauer) [inline catroid/src/main/java/org/catrobat/catroid/utils/FlashUtil.java]: Why was if (!CameraManager.getInstance().hasCurrentCameraFlash()) not enough?

File: catroid/src/test/java/org/catrobat/catroid/test/utiltests/FlashUtilTest.java [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+
+	@Before
+	public void setUp() {
+		oldCameraManagerInstance = CameraManager.getInstance();
+	}
+
+	@After
+	public void resetInstance() {
+		CameraManager.setInstance(oldCameraManagerInstance);
+	}
```
Comment (rschluesselbauer) [inline catroid/src/test/java/org/catrobat/catroid/test/utiltests/FlashUtilTest.java]: Setup & cleanup methods are unnecessary here. If other tests rely on CameraManager having a specific state, then they need to make sure it is in that state in their own tests.

File: catroid/src/main/java/org/catrobat/catroid/camera/CameraManager.java [hunk 5]
```diff
@@ -543,18 +543,19 @@ public void destroyStage() {
 		}
 	}
 
-	public boolean switchToCameraWithFlash() {
+	public boolean isCameraFlashAvailable() {
+		return (frontCameraInformation.flashAvailable || backCameraInformation.flashAvailable);
```
Comment (rschluesselbauer) [inline catroid/src/main/java/org/catrobat/catroid/camera/CameraManager.java]: If you have a device with less than 2 cameras you run into a NPE here, because frontCameraInformation / backCameraInformation would not have been initialized in the CameraManager constructor. (Happened to me on an emulated device without camera)

File: catroid/src/main/java/org/catrobat/catroid/camera/CameraManager.java [hunk 6]
```diff
@@ -543,18 +543,19 @@ public void destroyStage() {
...
+	}
+
+	public void switchToCameraWithFlash() {
 		if (hasCurrentCameraFlash()) {
-			return true;
+			return;
 		}
+
 		if (frontCameraInformation.flashAvailable) {
```
Comment (rschluesselbauer) [inline catroid/src/main/java/org/catrobat/catroid/camera/CameraManager.java]: Same thing goes for this if-else-construct -> NPE if frontCameraInformation is not initialized (because no front camera exists).

File: catroid/src/main/java/org/catrobat/catroid/utils/FlashUtil.java (hunk 3 above)
Comment (rschluesselbauer) [reply catroid/src/main/java/org/catrobat/catroid/utils/FlashUtil.java]: At a second glance - this is far from ignoring the flash brick!
If the currently active camera does not have a flash, it results in us destroying the stage - IMO the ...destroyStage() part needs to go.

File: catroid/src/test/java/org/catrobat/catroid/test/utiltests/FlashUtilTest.java (hunk 4 above)
Comment (84n4n4) [reply catroid/src/test/java/org/catrobat/catroid/test/utiltests/FlashUtilTest.java]: NoNoNoNo!
Always reset everything your tests changed in any kind of persistant state.

this can have so many side effects, not because a follow up test could need the camera manager directly, but the camera manager has 122 usages across the project, and if any of those its getting really tricky to debug and locate the problem. (eg. the compute dialog in formula editor accesses cameraManager)

i just spent 2 hours last week with a bug like that. someone mocked stage listener in a test and didnt have a teardown, leading to another test 100 tests apart in the test suite to fail due to a null pointer exception.

Comment (rschluesselbauer) [reply catroid/src/test/java/org/catrobat/catroid/test/utiltests/FlashUtilTest.java]: Good point, didn't think about it that way...
Sorry for misinformation then!

File: catroid/src/main/java/org/catrobat/catroid/camera/CameraManager.java [hunk 7]
```diff
@@ -310,6 +310,20 @@ public boolean hasCurrentCameraFlash() {
 		return currentCameraInformation.flashAvailable;
 	}
 
+	public boolean hasBackCameraFlash() {
+		if (hasBackCamera()) {
```
Comment (thmq) [inline catroid/src/main/java/org/catrobat/catroid/camera/CameraManager.java]: you can use "early return" here:

```
return hasBackCamera() && backCameraInformation.flashAvailable:
```

Java evaluates the right hand side of the condition first, so even no null pointer exception can occur here.

File: catroid/src/main/java/org/catrobat/catroid/camera/CameraManager.java [hunk 8]
```diff
@@ -310,6 +310,20 @@ public boolean hasCurrentCameraFlash() {
...
+			return backCameraInformation.flashAvailable;
+		}
+		return false;
+	}
+
+	public boolean hasFrontCameraFlash() {
+		if (hasFrontCamera()) {
```
Comment (thmq) [inline catroid/src/main/java/org/catrobat/catroid/camera/CameraManager.java]: see comment above.

File: catroid/src/main/java/org/catrobat/catroid/camera/CameraManager.java [hunk 9]
```diff
@@ -96,6 +97,11 @@ public static CameraManager getInstance() {
 		return instance;
 	}
 
+	@VisibleForTesting(otherwise = VisibleForTesting.NONE)
```
Comment (thmq) [inline catroid/src/main/java/org/catrobat/catroid/camera/CameraManager.java]: :+1: nice one

File: catroid/src/main/java/org/catrobat/catroid/utils/FlashUtil.java [hunk 10]
```diff
@@ -51,8 +51,13 @@ public static boolean isOn() {
 		return currentFlashValue;
 	}
 
+	public static boolean isPaused() {
+		return paused;
+	}
+
 	public static void pauseFlash() {
-		Log.d(TAG, "pauseFlash");
+		Log.d(TAG, "pauseFlash()");
```
Comment (thmq) [inline catroid/src/main/java/org/catrobat/catroid/utils/FlashUtil.java]: just remove the log altogether, as this was clearly added for debugging purposes. 

I think it is just bad practice to use log for debugging instead of Android Studio's debugger. breakpoints ftw! :smile:

File: catroid/src/main/java/org/catrobat/catroid/utils/FlashUtil.java [hunk 11]
```diff
@@ -64,7 +69,8 @@ public static void pauseFlash() {
 
 	public static void resumeFlash() {
 		Log.d(TAG, "resumeFlash()");
```
Comment (thmq) [inline catroid/src/main/java/org/catrobat/catroid/utils/FlashUtil.java]: please remove the log (see other comment above).

File: catroid/src/main/java/org/catrobat/catroid/utils/FlashUtil.java [hunk 12]
```diff
@@ -94,7 +100,13 @@ public static void reset() {
 	public static void initializeFlash() {
 		Log.d(TAG, "initializeFlash()");
```
Comment (thmq) [inline catroid/src/main/java/org/catrobat/catroid/utils/FlashUtil.java]: please remove (as with other logs).

File: catroid/src/main/java/org/catrobat/catroid/utils/FlashUtil.java [hunk 13]
```diff
@@ -112,6 +124,7 @@ public static void initializeFlash() {
 
 	public static void flashOn() {
 		Log.d(TAG, "flashOn()");
```
Comment (thmq) [inline catroid/src/main/java/org/catrobat/catroid/utils/FlashUtil.java]: please remove log here too, while you're at it.

File: catroid/src/main/java/org/catrobat/catroid/utils/FlashUtil.java [hunk 14]
```diff
@@ -123,6 +136,11 @@ public static void flashOn() {
 
 	public static void flashOff() {
 		Log.d(TAG, "flashOff()");
```
Comment (thmq) [inline catroid/src/main/java/org/catrobat/catroid/utils/FlashUtil.java]: another log that can be removed (I know it's not part of your commit, but please do it anyway).

File: catroid/src/main/java/org/catrobat/catroid/utils/FlashUtil.java (hunk 3 above)
Comment (DinosaurierRex) [reply catroid/src/main/java/org/catrobat/catroid/utils/FlashUtil.java]: As in the Ticket Description says
> Expected Behavior: The Flash brick should simply be ignored on devices without a flash.
i think we shouldn't destroy the stage. There should just nothing happen if you try to activate flash.
